File: plugins/dotnet/robocode.dotnet.tests.net/src/RbSerializerTest.cs
using System;
using System.IO;
using net.sf.robocode.api;
using net.sf.robocode.serialization;
using NUnit.Framework;
using robocode.dotnet.core.net.peer;
using robocode.util;

namespace robocode.dotnet.test.net
{
    [TestFixture]
    public class RbSerializerTest
    {
        private readonly byte[] ba;

        public RbSerializerTest()
        {
            var b = new sbyte[]
                        {
                            -95, -34, -34, -64, 0, 2, 7, 1, -81, 0, 0, 0, 1, 14, 45, -78, -99, -17, -61, 98, 64, 0, 0, 0
                            , 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, -116, 41, 41,
                            -1, -116, 41, 41, -1, -116, 41, 41, -1, -1, 0, 0, -1, -1, -1, -1, -1, -121, 68, -25, 74, 24,
                            87, -58, 63, 0, 0, 0, 0, 0, 0, 32, 64, 0, 0, 0, 0, -1, -1, -1, -1, -1, -1, -1, -1, 2, 0, 0,
                            0, 0, 0, 0, -16, 63, 1, 116, -75, 21, -5, -53, -18, -19, 63, 11, 0, 0, 0, 2, 0, 0, 0, 0, 0,
                            0, -16, 63, 0, -58, 109, 52, -128, -73, 64, -18, 63, 12, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, -16,
                            63, 1, 24, 38, 83, 5, -93, -110, -18, 63, -128, -1, -1, -1, -128, -128, 4, 4, 0, 0, 0, 116,
                            101, 115, 116, 8, 0, 0, 0, 116, 101, 115, 116, 68, 97, 116, 97, -128
                        };
            ba = new byte[b.Length];
            Buffer.BlockCopy(b, 0, ba, 0, b.Length);
            Module.Init();
            core.net.Module.Init();
        }

        [Test]
        public void empty()
        {
            //throw new ApplicationException();
            var ec = new ExecCommands();

            ec.setBodyTurnRemaining(150.123);
            ec.setTryingToPaint(true);
            var rbs = new RbSerializer();

            var outp = new MemoryStream(2048);

            int len = rbs.serialize(outp, RbSerializer.ExecCommands_TYPE, ec);
            byte[] buffer = outp.GetBuffer();
            var inp = new MemoryStream(buffer, 0, len);
            var ec2 = (ExecCommands) rbs.deserialize(inp);

            assertNear(ec2.getBodyTurnRemaining(), ec.getBodyTurnRemaining());
            Assert.AreEqual(ec2.IsTryingToPaint(), true);
        }

        [Test]
        public void withBullets()
        {
            var ec = new ExecCommands();

            ec.setBodyTurnRemaining(150.123);
            ec.getBullets().Add(new BulletCommand(1.0, true, 0.9354, 11));
            ec.getBullets().Add(new BulletCommand(1.0, false, 0.9454, 12));
            ec.getBullets().Add(new BulletCommand(1.0, true, 0.9554, -128));

            var outp = new MemoryStream(1024);
            var rbs = new RbSerializer();

            int len = rbs.serialize(outp, RbSerializer.ExecCommands_TYPE, ec);
            var inp = new MemoryStream(outp.GetBuffer(), 0, len);
            var ec2 = (ExecCommands) rbs.deserialize(inp);

            assertNear(ec2.getBodyTurnRemaining(), ec.getBodyTurnRemaining());
            assertNear(ec2.getBullets()[0].getPower(), 1.0);
            Assert.AreEqual(ec2.getBullets()[1].isFireAssistValid(), false);
            Assert.AreEqual(ec2.getBullets()[2].isFireAssistValid(), true);
            Assert.AreEqual(ec2.getBullets()[2].getBulletId(), -128);
        }

        [Test]
        public void withMessages()
        {
            var ec = new ExecCommands();

            ec.setBodyTurnRemaining(150.123);
            ec.getBullets().Add(new BulletCommand(1.0, true, 0.9354, 11));
            var data = new byte[20];

            data[10] = 10;
            ec.getTeamMessages().Add(new TeamMessage("Foo", "Bar", data));
            ec.getTeamMessages().Add(new TeamMessage("Foo", "Bar", null));

            var outp = new MemoryStream(1024);
            var rbs = new RbSerializer();

            int len = rbs.serialize(outp, RbSerializer.ExecCommands_TYPE, ec);
            var inp = new MemoryStream(outp.GetBuffer(), 0, len);
            var ec2 = (ExecCommands) rbs.deserialize(inp);

            Assert.AreEqual(ec2.getTeamMessages()[0].message[0], 0);
            Assert.AreEqual(ec2.getTeamMessages()[0].message[10], 10);
            Assert.AreEqual(ec2.getTeamMessages()[0].sender, "Foo");
            Assert.AreEqual(ec2.getTeamMessages()[0].recipient, "Bar");
            Assert.AreEqual(ec2.getTeamMessages()[1].message, null);
        }

        [Test]
        public void withProperties()
        {
            var ec = new ExecCommands();

            ec.setBodyTurnRemaining(150.123);
            ec.getBullets().Add(new BulletCommand(1.0, true, 0.9354, 11));
            ec.getTeamMessages().Add(new TeamMessage("Foo", "Bar", null));
            ec.getDebugProperties().Add(
                new DebugProperty("UTF8 Native characters", "P¯Ìliö ûluùouËk˝ k˘Ú ˙pÏl Ô·belskÈ Ûdy"));

            var outp = new MemoryStream(1024);
            var rbs = new RbSerializer();

            int len = rbs.serialize(outp, RbSerializer.ExecCommands_TYPE, ec);
            var inp = new MemoryStream(outp.GetBuffer(),0,len);
            var ec2 = (ExecCommands) rbs.deserialize(inp);

            Assert.AreEqual(ec2.getDebugProperties()[0].getKey(), "UTF8 Native characters");
            Assert.AreEqual(ec2.getDebugProperties()[0].getValue(), "P¯Ìliö ûluùouËk˝ k˘Ú ˙pÏl Ô·belskÈ Ûdy");
        }

        public static void assertNear(double v1, double v2)
        {
            Assert.AreEqual(v1, v2, Utils.NEAR_DELTA);
        }

        [Test]
        public void Test1()
        {
            var rbs = new RbSerializer();
            var ec2 = (ExecCommands) rbs.deserialize(new MemoryStream(ba));

            assertNear(ec2.getBodyTurnRemaining(), 150.123);
            assertNear(ec2.getBullets()[0].getPower(), 1.0);
            Assert.AreEqual(ec2.getBullets()[1].isFireAssistValid(), false);
            Assert.AreEqual(ec2.getBullets()[2].isFireAssistValid(), true);
            Assert.AreEqual(ec2.getBullets()[2].getBulletId(), -128);
            Assert.AreEqual(ec2.getDebugProperties()[0].getValue(), "testData");
        }
    }
}

File: robocode.core/src/test/java/net/sf/robocode/serialization/RbSerializerTest.java
/**
 * Copyright (c) 2001-2019 Mathew A. Nelson and Robocode contributors
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * https://robocode.sourceforge.io/license/epl-v10.html
 */
package net.sf.robocode.serialization;


import net.sf.robocode.peer.BulletCommand;
import net.sf.robocode.peer.DebugProperty;
import net.sf.robocode.peer.ExecCommands;
import net.sf.robocode.peer.TeamMessage;
import net.sf.robocode.robotpaint.Graphics2DSerialized;
import net.sf.robocode.security.HiddenAccess;

import org.junit.AfterClass;
import org.junit.Assert;
import org.junit.BeforeClass;
import org.junit.Test;

import robocode.util.Utils;

import javax.swing.*;

import java.awt.*;
import java.awt.geom.*;
import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;


/**
 * @author Pavel Savara (original)
 */
public class RbSerializerTest {

	Exception exception = null;
	
	@BeforeClass
	public static void init() {
		// we need to switch off engine classloader for this test
		System.setProperty("NOSECURITY", "true");
		System.setProperty("WORKINGDIRECTORY", "target//test-classes");
		System.setProperty("TESTING", "true");
		HiddenAccess.initContainer();
	}

	@AfterClass
	public static void cleanup() {
		System.setProperty("NOSECURITY", "false");
	}

	@Test
	public void empty() throws IOException {
		HiddenAccess.initContainer();
		ExecCommands ec = new ExecCommands();

		ec.setBodyTurnRemaining(150.123);
		ec.setTryingToPaint(true);

		ByteArrayOutputStream out = new ByteArrayOutputStream(1024);
		RbSerializer rbs = new RbSerializer();

		rbs.serialize(out, RbSerializer.ExecCommands_TYPE, ec);
		ByteArrayInputStream in = new ByteArrayInputStream(out.toByteArray());
		ExecCommands ec2 = (ExecCommands) rbs.deserialize(in);

		assertNear(ec2.getBodyTurnRemaining(), ec.getBodyTurnRemaining());
		Assert.assertEquals(ec2.isTryingToPaint(), true);
	}

	@Test
	public void withBullets() throws IOException {
		ExecCommands ec = new ExecCommands();

		ec.setBodyTurnRemaining(150.123);
		ec.getBullets().add(new BulletCommand(1.0, true, 0.9354, 11));
		ec.getBullets().add(new BulletCommand(1.0, false, 0.9454, 12));
		ec.getBullets().add(new BulletCommand(1.0, true, 0.9554, -128));

		ByteArrayOutputStream out = new ByteArrayOutputStream(1024);
		RbSerializer rbs = new RbSerializer();

		rbs.serialize(out, RbSerializer.ExecCommands_TYPE, ec);
		ByteArrayInputStream in = new ByteArrayInputStream(out.toByteArray());
		ExecCommands ec2 = (ExecCommands) rbs.deserialize(in);

		assertNear(ec2.getBodyTurnRemaining(), ec.getBodyTurnRemaining());
		assertNear(ec2.getBullets().get(0).getPower(), 1.0);
		Assert.assertEquals(ec2.getBullets().get(1).isFireAssistValid(), false);
		Assert.assertEquals(ec2.getBullets().get(2).isFireAssistValid(), true);
		Assert.assertEquals(ec2.getBullets().get(2).getBulletId(), -128);
	}

	@Test
	public void withMessages() throws IOException {
		ExecCommands ec = new ExecCommands();

		ec.setBodyTurnRemaining(150.123);
		ec.getBullets().add(new BulletCommand(1.0, true, 0.9354, 11));
		final byte[] data = new byte[20];

		data[10] = 10;
		ec.getTeamMessages().add(new TeamMessage("Foo", "Bar", data));
		ec.getTeamMessages().add(new TeamMessage("Foo", "Bar", null));

		ByteArrayOutputStream out = new ByteArrayOutputStream(1024);
		RbSerializer rbs = new RbSerializer();

		rbs.serialize(out, RbSerializer.ExecCommands_TYPE, ec);
		ByteArrayInputStream in = new ByteArrayInputStream(out.toByteArray());
		ExecCommands ec2 = (ExecCommands) rbs.deserialize(in);

		Assert.assertEquals(ec2.getTeamMessages().get(0).message[0], 0);
		Assert.assertEquals(ec2.getTeamMessages().get(0).message[10], 10);
		Assert.assertEquals(ec2.getTeamMessages().get(0).sender, "Foo");
		Assert.assertEquals(ec2.getTeamMessages().get(0).recipient, "Bar");
		Assert.assertEquals(ec2.getTeamMessages().get(1).message, null);
	}

	@Test
	public void withProperties() throws IOException {
		ExecCommands ec = new ExecCommands();

		ec.setBodyTurnRemaining(150.123);
		ec.getBullets().add(new BulletCommand(1.0, true, 0.9354, 11));
		ec.getTeamMessages().add(new TeamMessage("Foo", "Bar", null));
		ec.getDebugProperties().add(
				new DebugProperty("UTF8 Native characters", "Pøíliš žlu?ouèký kùò úpìl ïábelské ódy"));

		ByteArrayOutputStream out = new ByteArrayOutputStream(1024);
		RbSerializer rbs = new RbSerializer();

		rbs.serialize(out, RbSerializer.ExecCommands_TYPE, ec);
		ByteArrayInputStream in = new ByteArrayInputStream(out.toByteArray());
		ExecCommands ec2 = (ExecCommands) rbs.deserialize(in);

		Assert.assertEquals(ec2.getDebugProperties().get(0).getKey(), "UTF8 Native characters");
		Assert.assertEquals(ec2.getDebugProperties().get(0).getValue(), "Pøíliš žlu?ouèký kùò úpìl ïábelské ódy");
	}

	// @Test
	// 14 seconds for 1000 000,
	// 15x faster
	public void speed() throws IOException {
		ExecCommands ec = new ExecCommands();

		ec.setBodyTurnRemaining(150.123);
		ec.getBullets().add(new BulletCommand(1.0, true, 0.9354, 11));
		ec.getBullets().add(new BulletCommand(1.0, true, 0.9354, 11));
		ec.getBullets().add(new BulletCommand(1.0, true, 0.9354, 11));
		ec.getTeamMessages().add(new TeamMessage("Foo", "Bar", null));
		ec.getDebugProperties().add(new DebugProperty("ooooh", "aaaah"));

		ByteArrayOutputStream out = new ByteArrayOutputStream(1024);

		for (int i = 0; i < 1000000; i++) {
			out.reset();
			ec.setGunColor(i);
			RbSerializer rbs = new RbSerializer();

			rbs.serialize(out, RbSerializer.ExecCommands_TYPE, ec);
			ByteArrayInputStream in = new ByteArrayInputStream(out.toByteArray());
			ExecCommands ec2 = (ExecCommands) rbs.deserialize(in);

			Assert.assertEquals(ec2.getGunColor(), i);
		}
	}

	// @Test
	// 21 seconds for 100 000
	public void speed2() {
		ExecCommands ec = new ExecCommands();

		ec.setBodyTurnRemaining(150.123);
		ec.getBullets().add(new BulletCommand(1.0, true, 0.9354, 11));
		ec.getBullets().add(new BulletCommand(1.0, true, 0.9354, 11));
		ec.getBullets().add(new BulletCommand(1.0, true, 0.9354, 11));
		ec.getTeamMessages().add(new TeamMessage("Foo", "Bar", null));
		ec.getDebugProperties().add(new DebugProperty("ooooh", "aaaah"));

		ByteArrayOutputStream out = new ByteArrayOutputStream(1024);

		for (int i = 0; i < 100000; i++) {
			out.reset();
			ec.setGunColor(i);
			ExecCommands ec2 = (ExecCommands) ObjectCloner.deepCopy(ec);

			Assert.assertEquals(ec2.getGunColor(), i);
		}
	}

	@SuppressWarnings("serial")
	@Test
	public void graphics() throws InterruptedException {
		final Graphics2DSerialized sg = new Graphics2DSerialized();

		sg.setPaintingEnabled(true);
		sg.setBackground(Color.GREEN);
		sg.setColor(Color.RED);
		Arc2D a = new Arc2D.Double(Arc2D.PIE);

		a.setAngleExtent(10);
		a.setAngleStart(-30);
		a.setFrame(0, 0, 80, 80);
		sg.draw(a);

		sg.setColor(Color.BLUE);
		sg.draw(new Line2D.Double(99, 98, 78, 3));

		sg.setColor(Color.YELLOW);
		sg.draw(new Rectangle2D.Double(20, 20, 30, 50));

		sg.setColor(Color.BLACK);
		sg.drawLine(99, 3, 78, 3);
		sg.drawRect(90, 20, 30, 50);

		sg.setColor(Color.CYAN);

		sg.setStroke(new BasicStroke(1, 2, BasicStroke.JOIN_ROUND, 4, null, 0));
		sg.fill(new Rectangle2D.Double(20, 70, 30, 50));
		sg.fill(new Ellipse2D.Double(70, 70, 30, 50));

		sg.setColor(Color.MAGENTA);
		sg.fill(new RoundRectangle2D.Double(110, 70, 30, 50, 13.5, 16.1));

		exception = null;
		
		Canvas d = new Canvas() {
			@Override
			public void paint(Graphics g) {
				synchronized (this) {
					try {
						sg.processTo((Graphics2D) g);
					} catch (Exception e) {
						exception = e;
					}
				}
			}
		};

		d.setSize(200, 200);
		if (!java.awt.GraphicsEnvironment.isHeadless()) {
			JFrame f = new JFrame() {
			};

			f.add(d);
			f.pack();
			f.setVisible(true);
			f.setFocusable(true);
			Thread.sleep(100);
			f.setVisible(false);
		}

		Assert.assertNull("Exception occured: " + exception, exception);
	}

	public static void assertNear(double v1, double v2) {
		org.junit.Assert.assertEquals(v1, v2, Utils.NEAR_DELTA);
	}
}
